spec/integration/atomic_switcher_spec.rb

```diff
@@ -14,9 +14,105 @@
 
   describe 'switching' do
     before(:each) do
-      @origin      = table_create('origin')
-      @destination = table_create('destination')
+      @origin      = table_create("origin")
+      @destination = table_create("destination")
       @migration   = Lhm::Migration.new(@origin, @destination)
+      @connection.execute("SET GLOBAL innodb_lock_wait_timeout=3")
+      @connection.execute("SET GLOBAL lock_wait_timeout=3")
+    end
+
+   it "should retry on lock wait timeouts" do
+     begin
+       old_verbose, $VERBOSE = $VERBOSE, nil
+       old_retry_sleep = Lhm::AtomicSwitcher::RETRY_SLEEP_TIME
+       Lhm::AtomicSwitcher::RETRY_SLEEP_TIME = 0.2
+
+       mutex = Mutex.new
+       cond = ConditionVariable.new
+
+       locking_thread = Thread.new do 
+         with_per_thread_lhm_connection do |conn|
+
+           conn.sql('BEGIN')
+           conn.sql("DELETE from #{@destination.name}")
+           mutex.synchronize { sleep(1); cond.signal }
+           sleep(10)
+           conn.sql('ROLLBACK')
+         end
+       end
+
+       switching_thread = Thread.new do
+         with_per_thread_lhm_connection do |conn|
+           switcher = Lhm::AtomicSwitcher.new(@migration, conn)
+           mutex.synchronize do
+             cond.wait(mutex)
+             switcher.run
+           end
+           Thread.current[:retries] = switcher.retries
+         end
+       end
+
+       switching_thread.join
+       assert switching_thread[:retries] > 0, "The switcher did not retry"
+     ensure
+       Lhm::AtomicSwitcher::RETRY_SLEEP_TIME = old_retry_sleep
+       $VERBOSE = old_verbose
+     end
+   end
+
+   it "should give up on lock wait timeouts after MAX_RETRIES" do
+     begin
+       old_verbose, $VERBOSE = $VERBOSE, nil
+
+       old_retry_sleep = Lhm::AtomicSwitcher::RETRY_SLEEP_TIME
+       old_max_retries = Lhm::AtomicSwitcher::MAX_RETRIES
+       Lhm::AtomicSwitcher::RETRY_SLEEP_TIME = 0
+       Lhm::AtomicSwitcher::MAX_RETRIES = 2
```

Comment:
I hate having to change constants in here =( , maybe the api should use local ivars that use the constants as default, but you can change them?!?